glommio/src/sync/semaphore.rs

```diff
@@ -7,49 +7,115 @@ use std::cell::RefCell;
 use std::future::Future;
 use std::io::{Error, ErrorKind, Result};
 use std::pin::Pin;
-use std::rc::Rc;
 use std::task::{Context, Poll, Waker};
 
-use intrusive_collections::intrusive_adapter;
+use intrusive_collections::linked_list::LinkOps;
+use intrusive_collections::{container_of, offset_of, Adapter, PointerOps};
 use intrusive_collections::{LinkedList, LinkedListLink};
+use std::marker::PhantomPinned;
+use std::ptr::NonNull;
 
 #[derive(Debug)]
-struct Waiter {
-    node: Rc<WaiterNode>,
-    sem_state: Rc<RefCell<SemaphoreState>>,
+struct Waiter<'a> {
+    node: WaiterNode,
+    semaphore: &'a Semaphore,
 }
 
 #[derive(Debug)]
 struct WaiterNode {
     link: LinkedListLink,
-    state: RefCell<WaiterState>,
+    units: u64,
+    waker: RefCell<Option<Waker>>,
+
+    //waiter node can not be Unpin so its pointer could be used inside of intrusive
+    //collection, it also can not outlive the container which is guaranteed by the
+    //Waiter lifetime bound to the Semaphore which is container of all Waiters.
+    _p: PhantomPinned,
 }
 
-#[derive(Debug)]
-struct WaiterState {
-    units: u64,
-    waker: Option<Waker>,
+struct WaiterPointerOps;
+
+unsafe impl PointerOps for WaiterPointerOps {
+    type Value = WaiterNode;
+    type Pointer = NonNull<WaiterNode>;
+
+    unsafe fn from_raw(&self, value: *const Self::Value) -> Self::Pointer {
+        NonNull::new_unchecked(value as *mut Self::Value)
+    }
+
+    fn into_raw(&self, ptr: Self::Pointer) -> *const Self::Value {
+        ptr.as_ptr() as *const Self::Value
+    }
 }
 
-intrusive_adapter!(WaiterAdapter = Rc<WaiterNode> : WaiterNode {link : LinkedListLink});
+struct WaiterAdapter {
+    pointers_ops: WaiterPointerOps,
+    link_ops: LinkOps,
+}
+
+impl WaiterAdapter {
+    fn new() -> Self {
+        WaiterAdapter {
+            pointers_ops: WaiterPointerOps,
+            link_ops: LinkOps,
+        }
+    }
+}
+
+unsafe impl Adapter for WaiterAdapter {
```

Comment:
are those unsafe signatures mandated by the intrusive containers implementation ?

In general there is a lot of `unsafe` usage now which I definitely consider a disadvantage. I would like to spend some time to understand each of them better

Reply:
Hi @glommer , unsafe adapter trait is imposed by the intrusive collections. Unfortunately, all implementations which want to use pinning property of the Future and to avoid allocation require a good portion of unsafe code.

But if you check it, the scope of its usage fails in two categories only.
1. Usage of pinned mutable reference. It is always unsafe because you have to be sure that you do not break pinning guarantees. So you do not move the object which you use outside of Pin. That is why moved is wrapped into Pin and there are some islands of unsafe when mutable reference is needed.
2. Adapter which converts pointers from to pointer to the link and pointer to the field. (At previous version it was hidden by the macros)

All mutations are controlled in code using RefCell to do not break mutability guarantees. Only links to the next times are controlled by intrusive list. 

Also for your consideration bench of usage of semaphore before and after the changes (numbers are quite consistent):

before 

cost of acquiring uncontended semaphore: 14ns
cost of acquiring contended semaphore: 190ns

after

cost of acquiring uncontended semaphore: 12ns (-14%)
cost of acquiring contended semaphore: 164ns (-14%)

P.S. To ensure the code safety I will add the Drop implementation for the waiter to ensure that it is removed from the waiting list for sure. And will replace join3 using features_light.

Reply:
To be clear I am very pro using this, and if we have to mark them as unsafe , so be it.
The only change I am requiring at this time was about the `join`.

My comment on `unsafe` is just meant to set expectations: I may take a bit longer to merge this because I want to understand the usage a bit better (I'll ask questions if needed).